### PR TITLE
feat: FeeRebate: implementation of allowed callers

### PR DESCRIPTION
**Pull request by davidprepo**
https://www.notion.so/prepo/Fee-Rebate-SC-Spec-5723bb91a17c42e2a409ffd9af591085#4377633e1aae467d969d139e6dc0c518

Parent: https://github.com/prepo-io/prepo-monorepo/pull/594

### CODE REVIEW

File: apps/smart-contracts/core/test/AllowedCallers.test.ts [hunk 1]
```diff
@@ -8,32 +8,72 @@ import { AllowedCallers } from '../typechain'
...
+    it('reverts when callers.length != allowed.length', async () => {
+      await expect(
+        allowedCallers.setAllowedCallers([user.address], [true, false])
+      ).to.be.revertedWith('_callers.length != _allowed.length')
+    })
 
-    it('sets a single account as allowed caller', async () => {})
+    it('sets a single account as allowed caller', async () => {
+      await allowedCallers.setAllowedCallers([user.address], [true])
+      expect(await allowedCallers.isCallerAllowed(user.address)).to.be.true
```
Comment (ramenforbreakfast) [inline apps/smart-contracts/core/test/AllowedCallers.test.ts:36]:
```suggestion
      await allowedCallers.setAllowedCallers([user.address], [true])
      
      expect(await allowedCallers.isCallerAllowed(user.address)).to.be.true
```
New line in-between **act** and **assert**. Please do this for testcases below.

File: apps/smart-contracts/core/test/AllowedCallers.test.ts [hunk 2]
```diff
@@ -8,32 +8,72 @@ import { AllowedCallers } from '../typechain'
...
+    })
 
-    it('sets multiple accounts as allowed', async () => {})
+    it('sets two accounts as allowed', async () => {
```
Comment (ramenforbreakfast) [inline apps/smart-contracts/core/test/AllowedCallers.test.ts:39]: Should add expect statements at the top verifying `user` and `user2` are not allowed.

File: apps/smart-contracts/core/test/AllowedCallers.test.ts [hunk 3]
```diff
@@ -8,32 +8,72 @@ import { AllowedCallers } from '../typechain'
...
   describe('# setAllowedCallers', () => {
-    it('reverts if callers.length != allowed.length', async () => {})
+    it('reverts when callers.length != allowed.length', async () => {
+      await expect(
+        allowedCallers.setAllowedCallers([user.address], [true, false])
+      ).to.be.revertedWith('_callers.length != _allowed.length')
+    })
 
-    it('sets a single account as allowed caller', async () => {})
+    it('sets a single account as allowed caller', async () => {
```
Comment (ramenforbreakfast) [inline apps/smart-contracts/core/test/AllowedCallers.test.ts:34]: Should add statement at top verifying `user` is not allowed. Remember to include newline between this statement (**Arrange** block) and the **Act** block

**Review by ramenforbreakfast — APPROVED** [review]
LGTM, besides some testcase changes.

**Review by XavierEkkel — COMMENTED** [review]
please add test cases / adjust test names as per skeleton PR

File: apps/smart-contracts/core/test/AllowedCallers.test.ts [hunk 4]
```diff
@@ -11,36 +11,76 @@ describe('=> AllowedCallers', () => {
...
   })
 
   describe('initial state', () => {
-    it('has no allowed accounts', async () => {})
+    it('has no allowed accounts', async () => {
+      expect(await allowedCallers.isCallerAllowed(deployer.address)).to.be.false
+    })
   })
 
   describe('# setAllowedCallers', () => {
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/AllowedCallers.test.ts:27]: add `reverts if empty callers array` and `reverts if empty allowed array` test cases

File: apps/smart-contracts/core/test/AllowedCallers.test.ts [hunk 5]
```diff
@@ -11,37 +11,129 @@ describe('=> AllowedCallers', () => {
...
-    it('is idempotent for allowed account and disallowed account', async () => {})
+      await allowedCallers.setAllowedCallers([user.address, user2.address], [true, true])
 
-    it('emits AllowedCallersChange once for single account', async () => {})
+      expect(await allowedCallers.isCallerAllowed(user.address)).to.be.true
+      expect(await allowedCallers.isCallerAllowed(user2.address)).to.be.true
+    })
 
-    it('emits AllowedCallersChange once for multiple accounts', async () => {})
+    it('unsets a single account as allowed', async () => {
```
Comment (ramenforbreakfast) [inline apps/smart-contracts/core/test/AllowedCallers.test.ts:66]:
```suggestion
    it('sets an account as disallowed', async () => {
```

File: apps/smart-contracts/core/test/AllowedCallers.test.ts [hunk 6]
```diff
@@ -11,37 +11,129 @@ describe('=> AllowedCallers', () => {
...
 
-    it('unsets previously allowed account and sets new account', async () => {})
+      await allowedCallers.setAllowedCallers([user.address], [true])
 
-    it('is idempotent for single account', async () => {})
+      expect(await allowedCallers.isCallerAllowed(user.address)).to.be.true
+    })
 
-    it('is idempotent for multiple accounts', async () => {})
+    it('sets two accounts as allowed', async () => {
```
Comment (ramenforbreakfast) [inline apps/smart-contracts/core/test/AllowedCallers.test.ts:56]:
```suggestion
    it('allows multiple accounts', async () => {
```

File: apps/smart-contracts/core/test/AllowedCallers.test.ts [hunk 7]
```diff
@@ -11,37 +11,129 @@ describe('=> AllowedCallers', () => {
...
+      )
+    })
 
-    it('sets single account as disallowed', async () => {})
+    it('reverts when callers.length = 0 and allowed.length = 0', async () => {
+      await expect(allowedCallers.setAllowedCallers([], [])).to.be.revertedWith('empty array')
+    })
 
-    it('sets multiple accounts as disallowed', async () => {})
+    it('sets a single account as allowed caller', async () => {
```
Comment (ramenforbreakfast) [inline apps/smart-contracts/core/test/AllowedCallers.test.ts:48]:
```suggestion
    it('sets account as allowed', async () => {
```

File: apps/smart-contracts/core/test/AllowedCallers.test.ts [hunk 8]
```diff
@@ -11,37 +11,129 @@ describe('=> AllowedCallers', () => {
...
   describe('initial state', () => {
-    it('has no allowed accounts', async () => {})
+    it('has no allowed accounts', async () => {
+      expect(await allowedCallers.isCallerAllowed(deployer.address)).to.be.false
+    })
   })
 
   describe('# setAllowedCallers', () => {
-    it('reverts if array length mismatch', async () => {})
+    it('reverts when callers.length != allowed.length', async () => {
```
Comment (ramenforbreakfast) [inline apps/smart-contracts/core/test/AllowedCallers.test.ts:28]:
```suggestion
    it('reverts if array length mismatch', async () => {
```
Please follow the testcase convention already commented on here => https://github.com/prepo-io/prepo-monorepo/pull/571

I'm not going to comment on everyone, but you get the gist.

File: apps/smart-contracts/core/test/AllowedCallers.test.ts [hunk 9]
```diff
@@ -11,37 +11,129 @@ describe('=> AllowedCallers', () => {
...
+      await expect(
+        allowedCallers.setAllowedCallers([user.address], [true, false])
+      ).to.be.revertedWith('_callers.length != _allowed.length')
+    })
 
-    it('sets single account as allowed', async () => {})
+    it('reverts when callers.length = 0', async () => {
```
Comment (ramenforbreakfast) [inline apps/smart-contracts/core/test/AllowedCallers.test.ts:34]:
```suggestion
    it('reverts if caller array empty', async () => {
```
Similar to above comment, also remember to not use **when** but **if**

File: apps/smart-contracts/core/test/AllowedCallers.test.ts [hunk 10]
```diff
@@ -11,37 +11,129 @@ describe('=> AllowedCallers', () => {
...
+    it('unsets a single account as allowed', async () => {
+      await allowedCallers.setAllowedCallers([user.address], [true])
+      expect(await allowedCallers.isCallerAllowed(user.address)).to.be.true
+
+      await allowedCallers.setAllowedCallers([user.address], [false])
+
+      expect(await allowedCallers.isCallerAllowed(user.address)).to.be.false
+    })
+
+    it('unsets multiple accounts as allowed', async () => {
```
Comment (ramenforbreakfast) [inline apps/smart-contracts/core/test/AllowedCallers.test.ts:75]:
```suggestion
    it('sets multiple accounts as disallowed', async () => {
```

File: apps/smart-contracts/core/test/AllowedCallers.test.ts [hunk 11]
```diff
@@ -11,37 +11,129 @@ describe('=> AllowedCallers', () => {
...
 
-    it('unsets previously allowed account and sets new account', async () => {})
+      await allowedCallers.setAllowedCallers([user.address], [true])
 
-    it('is idempotent for single account', async () => {})
+      expect(await allowedCallers.isCallerAllowed(user.address)).to.be.true
+    })
 
-    it('is idempotent for multiple accounts', async () => {})
+    it('allows multiple accounts', async () => {
```
Comment (ramenforbreakfast) [inline apps/smart-contracts/core/test/AllowedCallers.test.ts:56]:
```suggestion
    it('sets multiple accounts as allowed', async () => {
```

File: apps/smart-contracts/core/test/AllowedCallers.test.ts [hunk 12]
```diff
@@ -11,37 +11,129 @@ describe('=> AllowedCallers', () => {
...
-    it('is idempotent for allowed account and disallowed account', async () => {})
+      await allowedCallers.setAllowedCallers([user.address, user2.address], [true, true])
 
-    it('emits AllowedCallersChange once for single account', async () => {})
+      expect(await allowedCallers.isCallerAllowed(user.address)).to.be.true
+      expect(await allowedCallers.isCallerAllowed(user2.address)).to.be.true
+    })
 
-    it('emits AllowedCallersChange once for multiple accounts', async () => {})
+    it('sets an account as disallowed', async () => {
```
Comment (ramenforbreakfast) [inline apps/smart-contracts/core/test/AllowedCallers.test.ts:66]:
```suggestion
    it('sets account as disallowed', async () => {
```

**Review by ramenforbreakfast — APPROVED** [review]
Re-reviewed, LGTM besides two naming nits.

File: apps/smart-contracts/core/contracts/AllowedCallers.sol [hunk 13]
```diff
@@ -6,17 +6,34 @@ import "./interfaces/IAllowedCallers.sol";
...
   function setAllowedCallers(address[] memory _callers, bool[] memory _allowed)
     external
     virtual
     override
-  {}
+  {
+    require(_callers.length > 0 && _allowed.length > 0, "empty array");
+    require(
+      _callers.length == _allowed.length,
+      "_callers.length != _allowed.length"
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/contracts/AllowedCallers.sol:22]:
```suggestion
      "Array length mismatch"
```

File: apps/smart-contracts/core/contracts/AllowedCallers.sol [hunk 14]
```diff
@@ -6,17 +6,34 @@ import "./interfaces/IAllowedCallers.sol";
...
+    _;
+  }
+
   function setAllowedCallers(address[] memory _callers, bool[] memory _allowed)
     external
     virtual
     override
-  {}
+  {
+    require(_callers.length > 0 && _allowed.length > 0, "empty array");
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/contracts/AllowedCallers.sol:19]:
```suggestion
    require(_callers.length > 0 && _allowed.length > 0, "Empty array");
```

File: apps/smart-contracts/core/test/AllowedCallers.test.ts [hunk 15]
```diff
@@ -11,37 +11,129 @@ describe('=> AllowedCallers', () => {
...
   describe('initial state', () => {
-    it('has no allowed accounts', async () => {})
+    it('has no allowed accounts', async () => {
+      expect(await allowedCallers.isCallerAllowed(deployer.address)).to.be.false
+    })
   })
 
   describe('# setAllowedCallers', () => {
-    it('reverts if array length mismatch', async () => {})
+    it('reverts if array length mistmatch', async () => {
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/AllowedCallers.test.ts:28]:
```suggestion
    it('reverts if array length mismatch', async () => {
```

File: apps/smart-contracts/core/test/AllowedCallers.test.ts [hunk 16]
```diff
@@ -11,37 +11,129 @@ describe('=> AllowedCallers', () => {
...
+      expect(await allowedCallers.isCallerAllowed(user.address)).to.be.true
+      expect(await allowedCallers.isCallerAllowed(user2.address)).to.be.true
+
+      await allowedCallers.setAllowedCallers([user.address, user2.address], [true, true])
+
+      expect(await allowedCallers.isCallerAllowed(user.address)).to.be.true
+      expect(await allowedCallers.isCallerAllowed(user2.address)).to.be.true
+    })
+
+    it('is idempotent for allowed account and disallowed account', async () => {
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/AllowedCallers.test.ts:116]:
```suggestion
    it('is idempotent if allowed account and disallowed account', async () => {
```

File: apps/smart-contracts/core/test/AllowedCallers.test.ts [hunk 17]
```diff
@@ -11,37 +11,129 @@ describe('=> AllowedCallers', () => {
...
+    it('is idempotent for an account', async () => {
+      await allowedCallers.setAllowedCallers([user.address], [true])
+      expect(await allowedCallers.isCallerAllowed(user.address)).to.be.true
+
+      await allowedCallers.setAllowedCallers([user.address], [true])
+
+      expect(await allowedCallers.isCallerAllowed(user.address)).to.be.true
+    })
+
+    it('is idempotent for multiple accounts', async () => {
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/AllowedCallers.test.ts:105]:
```suggestion
    it('is idempotent if multiple accounts', async () => {
```

File: apps/smart-contracts/core/test/AllowedCallers.test.ts [hunk 18]
```diff
@@ -11,37 +11,129 @@ describe('=> AllowedCallers', () => {
...
+      await allowedCallers.setAllowedCallers([user.address], [true])
+      expect(await allowedCallers.isCallerAllowed(user.address)).to.be.true
+
+      await allowedCallers.setAllowedCallers([user.address, user2.address], [false, true])
+
+      expect(await allowedCallers.isCallerAllowed(user.address)).to.be.false
+      expect(await allowedCallers.isCallerAllowed(user2.address)).to.be.true
+    })
+
+    it('is idempotent for an account', async () => {
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/AllowedCallers.test.ts:96]:
```suggestion
    it('is idempotent if single account', async () => {
```

File: apps/smart-contracts/core/test/AllowedCallers.test.ts [hunk 19]
```diff
@@ -11,37 +11,129 @@ describe('=> AllowedCallers', () => {
...
+
+      await allowedCallers.setAllowedCallers([user.address, user2.address], [false, false])
+
+      expect(await allowedCallers.isCallerAllowed(user.address)).to.be.false
+      expect(await allowedCallers.isCallerAllowed(user2.address)).to.be.false
+    })
+
+    it('disallows previously allowed account and allows new account', async () => {
+      await allowedCallers.setAllowedCallers([user.address], [true])
+      expect(await allowedCallers.isCallerAllowed(user.address)).to.be.true
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/AllowedCallers.test.ts:88]: also expect user2 is false

**Review by XavierEkkel — APPROVED** [review]
LGTM after nits

**Pull request merged**